### DISCUSSION TITLE
EntityView filter with OR

### DISCUSSION THREAD

**Discussion by Pilpin** [Blazebit/blaze-persistence#1740] [Q&A] [answered]
Hello,

I have an entity and view with a `LocalDate end` property and I would like to be able to query views filtering on this property where it can either be null or be greater than or equal to a passed `LocalDate value`.

I am wondering if 
- it is at all possible to do this using the `@AttributeFilter(s)` annotations, 
- it would be possible to do with a custom filter extending `AttributeFilterProvider`
- I should do it in a class extending `ViewFilterProvider` like so
```
@Override
public <T extends WhereBuilder<T>> T apply(T whereBuilder) {
    var today = LocalDate.now();
    return whereBuilder
            .whereOr()
            .where("end").isNull()
            .where("end").ge(today)
            .endOr();
}
```
- I should just do this in the criteriaBuilder like so
```
CriteriaBuilder<SomethingEntity> cb = cbf.create(em, SomethingEntity.class);
cb.whereOr().where("end").isNull().where("end").ge(value).endOr();
var settings = EntityViewSetting.create(SomethingView.class, first, size);
PagedList<SomethingView> list = evm.applySetting(settings, cb).getResultList()
```

Thank you.

**Comment by beikov** [comment, marked as answer]
A custom `AttributeFilterProvider` is probably most appropriate. The code can be similar to the one you use for view filters, although if the filter value is always the same i.e. `now()`, you might want to consider using a view filter instead.

**Reply by Pilpin** [reply]
Thank you. 
To make sure I get it, would this be a good custom `AttributeFilterProvider` for this purpose ?
```
public class EndAttributeFilter<FilterValue>  extends AttributeFilterProvider<FilterValue> {
    private final LocalDate date;

    public EndAttributeFilter(Object value) {
        if (value == null) {
            throw new NullPointerException("value");
        }

        if (value instanceof LocalDate) {
            this.date = (LocalDate) value;
        } else if (value instanceof SubqueryProvider) {
            throw new IllegalArgumentException("Subqueries are not allowed for the EndAttributeFilter");
        } else {
            throw new IllegalArgumentException("value should be a LocalDate");
        }
    }

    @Override
    public <T extends WhereBuilder<T>> T apply(T whereBuilder, String attributeExpression) {
        return whereBuilder.whereOr().where(attributeExpression).isNull().where(attributeExpression).ge(date).endOr();
    }
}
```

**Reply by beikov** [reply]
Yes, that's correct :)
